src/biz/casbin/access_control.rs

```diff
@@ -0,0 +1,1009 @@
+  workspace::{
+    access_control::WorkspaceAccessControl,
+    member_listener::{WorkspaceMemberAction, WorkspaceMemberNotification},
+  },
+};
+use app_error::AppError;
+use database_entity::dto::{AFAccessLevel, AFRole};
+use realtime::collaborate::{CollabAccessControl, CollabUserId};
+
+use super::{
+  Action, ActionType, ObjectType, POLICY_FIELD_INDEX_ACTION, POLICY_FIELD_INDEX_OBJECT,
+  POLICY_FIELD_INDEX_USER,
+};
+
+/// Manages access control using Casbin.
+///
+/// Stores access control policies in the form `subject, object, role`
+/// where `subject` is `uid`, `object` is `oid`, and `role` is [AFAccessLevel] or [AFRole].
+///
+/// Roles are mapped to the corresponding actions that they are allowed to perform.
+/// `FullAccess` has write
+/// `FullAccess` has read
+///
+/// Access control requests are made in the form `subject, object, action`
+/// and will be evaluated against the policies and mappings stored,
+/// according to the model defined.
+pub struct CasbinAccessControl {
+  pg_pool: PgPool,
+  enforcer: Arc<RwLock<casbin::Enforcer>>,
+}
+
+impl Clone for CasbinAccessControl {
+  fn clone(&self) -> Self {
+    Self {
+      pg_pool: self.pg_pool.clone(),
+      enforcer: Arc::clone(&self.enforcer),
+    }
+  }
+}
+
+impl CasbinAccessControl {
+  pub fn new(
+    pg_pool: PgPool,
+    collab_listener: broadcast::Receiver<CollabMemberNotification>,
+    workspace_listener: broadcast::Receiver<WorkspaceMemberNotification>,
+    enforcer: casbin::Enforcer,
+  ) -> Self {
+    let enforcer = Arc::new(RwLock::new(enforcer));
+    spawn_listen_on_workspace_member_change(workspace_listener, enforcer.clone());
+    spawn_listen_on_collab_member_change(collab_listener, enforcer.clone());
+    Self { pg_pool, enforcer }
+  }
+  pub fn new_collab_access_control(&self) -> CasbinCollabAccessControl {
+    CasbinCollabAccessControl {
+      casbin_access_control: self.clone(),
+    }
+  }
+
+  pub fn new_workspace_access_control(&self) -> CasbinWorkspaceAccessControl {
+    CasbinWorkspaceAccessControl {
+      casbin_access_control: self.clone(),
+    }
+  }
+
+  /// Update permission for a user.
+  ///
+  /// [`ObjectType::Workspace`] has to be paired with [`ActionType::Role`],
+  /// [`ObjectType::Collab`] has to be paired with [`ActionType::Level`],
+  async fn update(
+    &self,
+    uid: &i64,
+    obj: &ObjectType<'_>,
+    act: &ActionType,
+  ) -> Result<bool, AppError> {
+    let (obj_id, action) = match (obj, act) {
+      (ObjectType::Workspace(_), ActionType::Role(role)) => {
+        Ok((obj.to_string(), i32::from(role.clone()).to_string()))
+      },
+      (ObjectType::Collab(_), ActionType::Level(level)) => {
+        Ok((obj.to_string(), i32::from(*level).to_string()))
+      },
+      _ => Err(AppError::Internal(anyhow!(
+        "invalid object type and action type combination: object={:?}, action={:?}",
+        obj,
+        act
+      ))),
+    }?;
+
+    let current_policy = self
```

Comment:
```rust
let mut enforcer = self.enforcer.write().await;
    if let Some(policy) = enforcer
        .get_filtered_policy(POLICY_FIELD_INDEX_OBJECT, vec![obj_id.clone()])
        .into_iter()
        .find(|p| p[POLICY_FIELD_INDEX_USER] == uid.to_string()) {
        enforcer
            .remove_policy(policy)
            .await
            .map_err(|e| AppError::Internal(anyhow!("casbin error: {e:?}")))?;
    }

    enforcer
        .add_policy(vec![uid.to_string(), obj_id, action])
        .await
        .map_err(|e| AppError::Internal(anyhow!("casbin error: {e:?}")))
}
```


Holding the write lock of the enforcer throughout the function provides several benefits:

1. Ensures the entire operation (removing an old policy and adding a new one) is atomic. This is crucial for maintaining consistency in access control rules.

2. Prevents race conditions in a multi-threaded environment where multiple threads might try to modify policies simultaneously.